crypto/thread/arch/thread_posix.c

```diff
@@ -120,18 +120,22 @@ int ossl_crypto_mutex_try_lock(CRYPTO_MUTEX *mutex)
 
 void ossl_crypto_mutex_lock(CRYPTO_MUTEX *mutex)
 {
+    int rc;
     pthread_mutex_t *mutex_p;
 
     mutex_p = (pthread_mutex_t *)mutex;
-    pthread_mutex_lock(mutex_p);
+    rc = pthread_mutex_lock(mutex_p);
+    OPENSSL_assert(rc == 0);
```

Comment:
`OPENSSL_assert` causes an abort even in production builds. This is highly controversial with various people holding opposing views on whether this is a good idea. At the moment we tend to avoid this and use plain `assert` or `ossl_assert` instead as applicable.

Reply:
We still don't have an assert policy: https://github.com/openssl/technical-policies/issues/49

Reply:
OPENSSL_assert() makes a locking failure an immediate DoS. However not having it and continuing as if the lock is taken can have much worse consequences.

Having checks for all mutex lock calls pollutes the code seriously and in most cases the locking failure indicates that there are much deeper problems with the state of the process that terminating is the best outcome anyway.

Choose your evil.

Reply:
> `OPENSSL_assert` causes an abort even in production builds. This is highly controversial with various people holding opposing views on whether this is a good idea. At the moment we tend to avoid this and use plain `assert` or `ossl_assert` instead as applicable.

This is a deliberate choice on my part.

If this assert is ever hit, we have a serious locking bug which is no less serious than a bug causing e.g. SIGSEGV. Leaving this only in debug builds sort of defeats the point of adding this assert.

Ultimately there are a lot of different things that can happen to make a process blow up if locking is mishandled (SIGSEGV, a lock call hanging forever, or this assertion) so it's a false premise to suppose avoiding `OPENSSL_assert` leads to some kind of greater "resilience". On the contrary I think this (`OPENSSL_assert`, not assert, not ossl_assert) is the only safe way to handle this eventuality.